### PR TITLE
fix 'We need to override the window.performance.getEntriesByType function'

### LINKED ISSUE #1214: We need to override the window.performance.getEntriesByType function
Resource with statistics fails with 500 status code on [avito.ru](https://www.avito.ru/).

Information getting with the `window.performance.getEntriesByType` function contains proxy url.
For example, function from avito:
```javascript
    function r() {
        if ("function" != typeof window.performance.getEntriesByType)
            return null;
        for (var e = [], n = window.performance.getEntriesByType("resource") || [], t = 0; t < __get$(n, "length"); t++) {
            var r = __get$(n, t)
              , o = r.name.match(v);
            o && e.push({
                origin: s(r.name),
                extension: o[1],
                size: {
                    transfer: r.transferSize,
                    encoded: r.encodedBodySize,
                    decoded: r.decodedBodySize
                },
                response: {
                    start: r.responseStart,
                    end: r.responseEnd
                }
            })
        }
        var a = window.performance.getEntriesByType("navigation");
        return a && __get$(a, "length") && (a = a[0],
        e.push({
            origin: w,
            extension: "html",
            size: {
                transfer: a.transferSize,
                encoded: a.encodedBodySize,
                decoded: a.decodedBodySize
            },
            response: {
                start: a.responseStart,
                end: a.responseEnd
            }
        })),
        __get$(e, "length") ? e : null
    }
```

### PR DESCRIPTION
**Pull request by Farfurix**
Add PerformanceEntry name property getter to native methods. Describes PerformanceNavigationTiming case for it. Part of #1214

### CODE REVIEW

File: src/client/sandbox/native-methods.js [hunk 1]
```diff
@@ -195,6 +195,13 @@ class NativeMethods {
 
         this.messageEventOriginGetter = win.Object.getOwnPropertyDescriptor(window.MessageEvent.prototype, 'origin').get;
 
+        if (win.PerformanceEntry) {
+            const namePropDescripor = win.Object.getOwnPropertyDescriptor(window.PerformanceEntry.prototype, 'name');
+
+            if (namePropDescripor && namePropDescripor.get && namePropDescripor.configurable)
```
Comment (miherlosev) [inline src/client/sandbox/native-methods.js]: Try to simplify as:
`this.performanceEntryNameGetter = win.Object.getOwnPropertyDescriptor(window.PerformanceEntry.prototype, 'name').get`

File: src/client/utils/dom.js [hunk 2]
```diff
@@ -628,6 +628,10 @@ export function isMessageEvent (e) {
     return instanceToString(e) === '[object MessageEvent]';
 }
 
+export function isPerformanceNavigationTiming (entry) {
```
Comment (LavrovArtem) [inline src/client/utils/dom.js]: Write test for this function.

File: src/client/sandbox/node/window.js [hunk 3]
```diff
@@ -525,6 +532,30 @@ export default class WindowSandbox extends SandboxBase {
         nativeMethods.objectDefineProperty
             .call(window.Object, window.MessageEvent.prototype, 'origin', originPropDescriptor);
 
+        if (window.PerformanceEntry) {
+            if (nativeMethods.performanceEntryNameGetter) {
+                const namePropDescriptor = nativeMethods.objectGetOwnPropertyDescriptor
```
Comment (LavrovArtem) [inline src/client/sandbox/node/window.js]: Rebase your branch on the `master` branch and use the [`overrideDescriptor`](https://github.com/DevExpress/testcafe-hammerhead/blob/master/src/client/utils/override-descriptor.js) function.

File: src/client/sandbox/node/window.js [hunk 4]
```diff
@@ -525,6 +532,30 @@ export default class WindowSandbox extends SandboxBase {
...
+                    .call(window.Object, window.PerformanceEntry.prototype, 'name');
+
+                namePropDescriptor.get = function () {
+                    const entry = this;
```
Comment (miherlosev) [inline src/client/sandbox/node/window.js]: Why do we need an additional variable?

File: src/client/sandbox/node/window.js (hunk 3 above)
Comment (miherlosev) [inline src/client/sandbox/node/window.js]: For overriding the property descriptor use the `overrideDescriptor` function - https://github.com/DevExpress/testcafe-hammerhead/blob/master/src/client/utils/override-descriptor.js

File: test/client/fixtures/sandbox/node/classes-test.js [hunk 5]
```diff
@@ -173,6 +173,23 @@ if (window.MutationObserver) {
     });
 }
 
+module('PerformanceEntry');
```
Comment (miherlosev) [inline test/client/fixtures/sandbox/node/classes-test.js]: the `module` directive needs for group a lot of tests.
I see here only one.
Remove the `module` directive and renamte test as `PerformanceNavigationTiming.name`.

File: src/client/sandbox/node/window.js [hunk 6]
```diff
@@ -525,6 +532,30 @@ export default class WindowSandbox extends SandboxBase {
...
+                    const url   = nativeMethods.performanceEntryNameGetter.call(this);
+
+                    if (isPerformanceNavigationTiming(entry)) {
```
Comment (miherlosev) [inline src/client/sandbox/node/window.js]: Add separate tests for `isPerformanceNavigationTiming` function.
For example see this https://github.com/DevExpress/testcafe-hammerhead/blob/6071663c7f72afe6bd74e72d6ab3b3628f9ce1ab/test/client/fixtures/utils/dom-test.js#L167

File: src/client/sandbox/node/window.js (hunk 4 above)
Comment (Farfurix) [reply src/client/sandbox/node/window.js]: okay, I'll delete it

File: src/client/sandbox/node/window.js [hunk 7]
```diff
@@ -542,6 +549,23 @@ export default class WindowSandbox extends SandboxBase {
             return nativeMethods.elementChildElementCountGetter.call(this);
         });
 
+        if (window.PerformanceEntry) {
```
Comment (miherlosev) [inline src/client/sandbox/node/window.js]: Unnecessary condition

File: test/client/fixtures/utils/dom-test.js [hunk 8]
```diff
@@ -184,6 +184,11 @@ test('isMessageEvent', function () {
         });
 });
 
+test('isPerformanceNavigationTiming', function () {
+    if (window.PerformanceEntry)
```
Comment (miherlosev) [inline test/client/fixtures/utils/dom-test.js]: Wrongly structured.

Replace with
```js
if (window.PerformanceEntry) {
   test('isPerformanceNavigationTiming', function () {
       ok(domUtils.isPerformanceNavigationTiming(window.performance.getEntriesByType('navigation');
      // add tests for 2 another PerformanceEntry types
   }
}
```

File: src/client/sandbox/node/window.js [hunk 9]
```diff
@@ -542,6 +549,21 @@ export default class WindowSandbox extends SandboxBase {
             return nativeMethods.elementChildElementCountGetter.call(this);
         });
 
+        if (nativeMethods.performanceEntryNameGetter) {
+            overrideDescriptor(window.PerformanceEntry.prototype, 'name', function () {
+                const url = nativeMethods.performanceEntryNameGetter.call(this);
```
Comment (miherlosev) [inline src/client/sandbox/node/window.js]: Let it be `name`

File: test/client/fixtures/sandbox/node/classes-test.js [hunk 10]
```diff
@@ -7,6 +7,21 @@ var INTERNAL_ATTRS  = hammerhead.get('../processing/dom/internal-attributes');
 var browserUtils  = hammerhead.utils.browser;
 var nativeMethods = hammerhead.nativeMethods;
 
+if (window.PerformanceEntry) {
+    test('name property of PerformanceNavigationTiming', function () {
```
Comment (miherlosev) [inline test/client/fixtures/sandbox/node/classes-test.js]: `test('PerformanceNavigationTiming.name')`

**Comment by Farfurix** [comment]
@testcafe-build-bot \retest

**Comment by Farfurix** [comment]
@testcafe-build-bot \retest

**Comment by Farfurix** [comment]
Part of #1214

File: src/client/sandbox/native-methods.js [hunk 11]
```diff
@@ -199,7 +199,8 @@ class NativeMethods {
         this.elementChildElementCountGetter = win.Object.getOwnPropertyDescriptor(window.Element.prototype, 'childElementCount').get;
 
         if (win.PerformanceEntry)
-            this.performanceEntryNameGetter = win.Object.getOwnPropertyDescriptor(window.PerformanceEntry.prototype, 'name').get;
+            if (win.PerformanceNavigationTiming)
```
Comment (miherlosev) [inline src/client/sandbox/native-methods.js]: I guess the only one condition `if (win.PerformanceNavigationTiming)` is enough.

Comment (Farfurix) [reply src/client/sandbox/native-methods.js]: Thanks, I found it out in documentation: "PerformanceEntry instances will always be one of the following subtypes: ..., PerformanceNavigationTiming, ... "

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Pull request merged**